### PR TITLE
Fix #609 - Implements HTTP Caching for HTML routes

### LINKED ISSUE #609: HTML resources HTTP Caching is broken on webcompat.com
We need to have a better story at caching HTML resources on webcompat.com.
See #590 and comments on #608

### PR DESCRIPTION
**Pull request by karlcow**
This currently doesn't fix all HTML resources served on our site, but just the issue view.
I would love to get it on staging, because of our nginx layer and see how it flies. Locally, things seem to go well.

```
→ nosetests -v

API access to comments greater than 30 returns pagination in Link ... ok
API issue for a non existent number returns JSON 404. ... ok
API issue search with bad keywords returns JSON 404. ... ok
API access to labels without auth returns JSON 200. ... ok
API with wrong parameter returns JSON 404. ... ok
API setting labels without auth returns JSON 403 error code. ... ok
API access to user activity without auth returns JSON 401. ... ok
API with wrong category returns JSON 404. ... ok
API with wrong route returns JSON 404. ... ok
test_domain_name (tests.test_form.TestForm) ... ok
Make sure wrap_metadata and get_metadata methods work. ... ok
test_normalize_url (tests.test_form.TestForm) ... ok
Test HTTP Links formating. ... ok
Test browser parsing via get_browser helper method. ... ok
Test browser name parsing via get_browser_name helper method. ... ok
Test OS parsing via get_os helper method. ... ok
Test that API params are correctly converted to Search API. ... ok
normalize_api_params shouldn't transform unknown params. ... ok
Test HTTP Links parsing for GitHub only. ... ok
test_rewrite_and_sanitize_link (tests.test_helpers.TestHelpers) ... ok
Test we're correctly rewriting the passed in link. ... ok
Test that we're removing access_token parameters. ... ok
Check Cache-Control for issues. ... ok
Check ETAG for issues. ... ok
Check if we receive a 304 Not Modified. ... ok
testBase64ScreenshotUploads (tests.test_uploads.TestUploads) ... ok
Test that /upload/ doesn't let you GET. ... ok
testRegularUploads (tests.test_uploads.TestUploads) ... ok
Test that /about exists. ... ok
Test that asks user to log in before displaying activity. ... ok
Test that the home page exists. ... ok
Test issues and integer for: ... ok
Test that the /issues/<number> exists, and does not redirect. ... ok
Test that the /issues route gets 200 and does not redirect. ... ok
Webhook related tests. ... ok
Test that the /login route 302s to GitHub. ... ok
Test that /issues/new exists. ... ok
Test that /privacy exists. ... ok
Rate Limit URI sends 200 OK and starts with Current user. ... ok
Test that the /tools/cssfixme route gets 200. ... ok
Test that the /tools/cssfixme route gets 200 with ?url query. ... ok
Test that the /tools/cssfixme route gets 200 with bad ?url query. ... ok

----------------------------------------------------------------------
Ran 42 tests in 9.196s

OK
```

And on the CLI.

First request:

```
→ http --print Hh GET http://localhost:5000/issues/100 
```

```http
GET /issues/100 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: localhost:5000
User-Agent: HTTPie/0.9.3
```

The 1st response

```http
HTTP/1.0 200 OK
Cache-Control: private, max-age=31536000
Content-Length: 9366
Content-Type: text/html; charset=utf-8
Date: Thu, 09 Feb 2017 04:46:16 GMT
ETag: "bce46f17692195b267de8026b3c5f66f"
Server: Werkzeug/0.10.4 Python/2.7.10
```

Then the second request with `If-None-Match`

```
→ http --print Hhb GET http://localhost:5000/issues/100 'If-None-Match:"bce46f17692195b267de8026b3c5f66f"'
```

```http
GET /issues/100 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: localhost:5000
If-None-Match: "bce46f17692195b267de8026b3c5f66f"
User-Agent: HTTPie/0.9.3
```

with the second response:

```http
HTTP/1.0 304 NOT MODIFIED
Cache-Control: private, max-age=31536000
Connection: close
Date: Thu, 09 Feb 2017 04:46:56 GMT
ETag: "bce46f17692195b267de8026b3c5f66f"
Server: Werkzeug/0.10.4 Python/2.7.10



```

### CODE REVIEW

**Comment by karlcow** [comment]
ok let's ask 
r? @miketaylr 
for a preliminary review.

File: webcompat/views.py [hunk 1]
```diff
@@ -207,7 +209,13 @@ def show_issue(number):
     if session.get('show_thanks'):
         flash(number, 'thanks')
         session.pop('show_thanks')
-    return render_template('issue.html', number=number)
+    content = render_template('issue.html', number=number)
+    response = make_response(content)
+    # We can cache for one year. Etag is based on the content.
+    response = set_cache_control(response, 31536000)
```
Comment (karlcow) [inline webcompat/views.py]: Ah! I might be a bit too agressive on the caching here.  I was wondering why the browser was not sending any conditional requests to the server

`Cache-Control: private, max-age=31536000`

which sets it to **1 year**. From the browser point of view it takes priority on the etag, and the browser will not even attempt to do `if-none-match` (aka conditional request). As long as the resource is in the cache and the 1 year has not passed, no attempt of revalidation will be done.

So basically, we need to decide what is the right balance in between:
* cache-control: Keeping the resources for x seconds. No hit to the server.
* etag: new etag when the resource has changed. A hit to the server with an eventual empty body and a 304. Or the full 200, if it has changed.

**Comment by miketaylr** [comment]
Let me deploy this to staging so you can play around with it. I won't have time for review until tomorrow.

**Comment by miketaylr** [comment]
🍪 for new tests!

**Comment by miketaylr** [comment]
@karlcow deployed to staging.

**Comment by karlcow** [comment]
Thanks @miketaylr 

https://staging.webcompat.com/issues/515

```http
HTTP/1.1 200 OK
Server: nginx/1.1.19
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Vary: Accept-Encoding
Cache-Control: private, max-age=31536000
Etag: "f9530168e98a2537ad066564ef2f38e5"
Date: Fri, 10 Feb 2017 01:09:30 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
x-xss-protection: 1; mode=block
x-frame-options: DENY
Content-Encoding: gzip
```

Click reload

Request
```http
GET /issues/515 HTTP/1.1
Host: staging.webcompat.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Referer: https://staging.webcompat.com/
Cookie: session=eyJjc3JmX3Rva2VuIjp7IiBiIjoiTXpjeVl6aGhNV0kwTW1Sa05tUXpaV0kzWlRrM05UQXdPRGc0WlRJNFlqZ3hOR00zTURReU13PT0ifX0.C36kNA.Z-UEcg07yZkalpEwnA9qivmYqO0; _ga=GA1.2.1918118046.1486688948; _gat=1
Authorization: Basic czpz
Connection: keep-alive
Upgrade-Insecure-Requests: 1
If-None-Match: "f9530168e98a2537ad066564ef2f38e5"
Cache-Control: max-age=0
```

Response
```http
HTTP/1.1 304 NOT MODIFIED
Server: nginx/1.1.19
Connection: keep-alive
Cache-Control: private, max-age=31536000
Etag: "f9530168e98a2537ad066564ef2f38e5"
Date: Fri, 10 Feb 2017 01:12:03 GMT
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
x-xss-protection: 1; mode=block
x-frame-options: DENY
```

So this is working as expected. \o/

**Comment by karlcow** [comment]
I'm changing a bit the strategy to have more flexbility for other resources/routes.

I created a decorator defining the cache_policy
It also means that in the future we will be able to play a bit more with it and do fine tuning.

The routes without caching

* `/`
* `/issues`
* `/issues/nnnn` DONE.
* `/about`
* `/privacy`
* `/contributors`
* `/me`
* `/activity/karlcow`

**Comment by karlcow** [comment]
Don't review yet :)

**Comment by miketaylr** [comment]
> Don't review yet :)

Deal!

File: webcompat/helpers.py [hunk 2]
```diff
@@ -456,3 +459,29 @@ def api_request(method, path, params=None, data=None):
...
     else:
         abort(404)
+
+
+def cache_policy(private=True, uri_max_age=3600):
+    '''Implements a HTTP Cache Decorator.
+
+    Adds Cache-Control headers.
+    Adds Etag based on HTTP Body.
+    Sends a 304 Not Modified in cas of If-None-Match.
```
Comment (miketaylr) [inline webcompat/helpers.py]: nit: s/cas/case/

**Comment by karlcow** [comment]
```
→ nosetests -v
/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask_limiter/extension.py:124: UserWarning: Use of the default `get_ipaddr` function is discouraged. Please refer to https://flask-limiter.readthedocs.org/#rate-limit-domain for the recommended configuration
  UserWarning
API access to comments greater than 30 returns pagination in Link ... ok
API issue for a non existent number returns JSON 404. ... ok
API issue search with bad keywords returns JSON 404. ... ok
API access to labels without auth returns JSON 200. ... ok
API with wrong parameter returns JSON 404. ... ok
API setting labels without auth returns JSON 403 error code. ... ok
API access to user activity without auth returns JSON 401. ... ok
API with wrong category returns JSON 404. ... ok
API with wrong route returns JSON 404. ... ok
test_domain_name (tests.test_form.TestForm) ... ok
Make sure wrap_metadata and get_metadata methods work. ... ok
test_normalize_url (tests.test_form.TestForm) ... ok
Test HTTP Links formating. ... ok
Test browser parsing via get_browser helper method. ... ok
Test browser name parsing via get_browser_name helper method. ... ok
Test OS parsing via get_os helper method. ... ok
Test that API params are correctly converted to Search API. ... ok
normalize_api_params shouldn't transform unknown params. ... ok
Test HTTP Links parsing for GitHub only. ... ok
test_rewrite_and_sanitize_link (tests.test_helpers.TestHelpers) ... ok
Test we're correctly rewriting the passed in link. ... ok
Test that we're removing access_token parameters. ... ok
Check Cache-Control for issues. ... ok
Check ETAG for issues. ... ok
Checks if we receive a 304 Not Modified. ... ok
testBase64ScreenshotUploads (tests.test_uploads.TestUploads) ... ok
Test that /upload/ doesn't let you GET. ... ok
testRegularUploads (tests.test_uploads.TestUploads) ... ok
Test that /about exists. ... ok
Test that asks user to log in before displaying activity. ... ok
Test that the home page exists. ... ok
Test issues and integer for: ... ok
Test that the /issues/<number> exists, and does not redirect. ... ok
Test that the /issues route gets 200 and does not redirect. ... ok
Webhook related tests. ... ok
Test that the /login route 302s to GitHub. ... ok
Test that /issues/new exists. ... ok
Test that /privacy exists. ... ok
Rate Limit URI sends 200 OK and starts with Current user. ... ok
Test that the /tools/cssfixme route gets 200. ... ok
Test that the /tools/cssfixme route gets 200 with ?url query. ... ok
Test that the /tools/cssfixme route gets 200 with bad ?url query. ... ok

----------------------------------------------------------------------
Ran 42 tests in 6.190s

OK
```

Let's start a formal review r? @miketaylr

File: webcompat/db/__init__.py [hunk 3]
```diff
@@ -45,6 +45,7 @@ def __init__(self, issue_id, summary, url, body):
         self.url = url
         self.body = body
 
+
```
Comment (miketaylr) [inline webcompat/db/__init__.py]: (a random whitespace change in the DB code feels funny ^__^)

File: webcompat/helpers.py [hunk 4]
```diff
@@ -456,3 +459,29 @@ def api_request(method, path, params=None, data=None):
                 get_response_headers(resource))
     else:
         abort(404)
+
+
+def cache_policy(private=True, uri_max_age=3600):
```
Comment (miketaylr) [inline webcompat/helpers.py]: Can we add a comment about what the default `uri_max_age=3600` represents? HTTP noobs like me have to look it up.

(The `private` arg has a nice inline comment)

File: webcompat/views.py [hunk 5]
```diff
@@ -201,14 +204,17 @@ def create_issue():
...
     '''Route to display a single issue.'''
     if g.user:
         get_user_info()
     if session.get('show_thanks'):
         flash(number, 'thanks')
         session.pop('show_thanks')
-    return render_template('issue.html', number=number)
+    content = render_template('issue.html', number=number)
+    response = make_response(content)
+    return response
```
Comment (miketaylr) [inline webcompat/views.py]: nit: Personally I would write `return make_response(content)` to avoid the extra variable here, but that's up to you.

File: webcompat/views.py [hunk 6]
```diff
@@ -143,6 +144,7 @@ def index():
 
 
 @app.route('/issues')
+@cache_policy(private=True, uri_max_age=86400)
```
Comment (miketaylr) [inline webcompat/views.py]: So far all instances of this decorator use `uri_max_age=86400` -- should that be the default, rather than 3600? Or do you have in mind using the decorator with the default somewhere else?

File: webcompat/db/__init__.py (hunk 3 above)
Comment (karlcow) [inline webcompat/db/__init__.py]: Probably I should have fixed it elsewhere. :)
It was a clean up of the syntax

File: webcompat/helpers.py (hunk 4 above)
Comment (karlcow) [inline webcompat/helpers.py]: good point.

File: webcompat/views.py (hunk 5 above)
Comment (karlcow) [inline webcompat/views.py]: yup it's a left over of something I was doing before.

File: webcompat/views.py (hunk 6 above)
Comment (karlcow) [inline webcompat/views.py]: Good point. Will do.

File: webcompat/views.py (hunk 5 above)
Comment (karlcow) [inline webcompat/views.py]: Ah! no… It's on purpose. For the decorator to be applied I need to have the make_response before.

**Comment by karlcow** [comment]
@miketaylr Ok I squashed everything into one commit. :) because rebasing with intertwined modifications was probably not a good idea. If you prefer I can submit separate commits for testing, decorator and views in a new branch. ^_^ No issue.

File: webcompat/views.py (hunk 5 above)
Comment (miketaylr) [inline webcompat/views.py]: Ah interesting. Did not know!

**Comment by miketaylr** [comment]
> @miketaylr Ok I squashed everything into one commit. :) because rebasing with intertwined modifications was probably not a good idea. If you prefer I can submit separate commits for testing, decorator and views in a new branch. ^_^ No issue.

That's fine too. Doing gigantic interactive rebases can easily get weird.

File: webcompat/views.py (hunk 6 above)
Comment (miketaylr) [inline webcompat/views.py]: Would you like to do this before merge, or in a follow up commit?

**Comment by miketaylr** [comment]
r=me, assuming you'll make the `uri_max_age=86400` default change. Feel free to merge and add the "add to changelog" label after that -- thanks!

File: webcompat/views.py (hunk 6 above)
Comment (karlcow) [inline webcompat/views.py]: hmmm? :) I did change it already ;)
See https://github.com/webcompat/webcompat.com/pull/1335/files#diff-23e0a36bbee957147173901eb77913cbR463

**Comment by karlcow** [comment]
> r=me, assuming you'll make the uri_max_age=86400 default change. Feel free to merge and add the "add to changelog" label after that -- thanks!

It was done no? So assuming that @miketaylr had missed the change
https://github.com/webcompat/webcompat.com/pull/1335/files/bab6f6df1676b65f891ace94527d4596f176dcbf#diff-23e0a36bbee957147173901eb77913cbR463

I'll move forward.

Thanks for the review! Very helpful.

**Comment by miketaylr** [comment]
> It was done no? So assuming that @miketaylr had missed the change

I wasn't even looking -- I assumed you would merge it after you made the change. 😄 

Happy this landed!
